src/validation.cpp

```diff
@@ -1090,6 +1090,33 @@ bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTrans
     return AcceptToMemoryPoolWithTime(chainparams, pool, state, tx, GetTime(), plTxnReplaced, bypass_limits, test_accept, fee_out);
 }
 
+// Calculate the fee of a transaction by looking up inputs from the mempool if
+// necessary. Return true if the fee was successfully calculated, or false if
+// an error with the transaction was discovered.
+// If the transaction is valid, then the fee returned is correct.
+// However, success does not imply that the transaction is valid.
+bool GetTransactionFee(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx, CAmount *fee_out)
```

Comment:
I think an `Optional<CAmount>` is the most natural return type here, rather than a 'success' bool and an out-param. Unset would mean "transaction invalid" and set would mean "fee calculated (but transaction not necessarily valid)".